go/cmd/pgctld/command/server.go

```diff
@@ -0,0 +1,355 @@
+/*
+Copyright 2025 The Multigres Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package command
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"net"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"google.golang.org/grpc"
+
+	pb "github.com/multigres/multigres/go/pb"
+)
+
+func init() {
+	Root.AddCommand(serverCmd)
+}
+
+var serverCmd = &cobra.Command{
+	Use:   "server",
+	Short: "Run pgctld as a gRPC server daemon",
+	Long:  `Run pgctld as a background gRPC server daemon to handle PostgreSQL management requests.`,
+	RunE:  runServer,
+}
+
+func runServer(cmd *cobra.Command, args []string) error {
+	logger := slog.Default()
+
+	port := viper.GetInt("grpc-port")
+
+	// Create gRPC server
+	grpcServer := grpc.NewServer()
+
+	// Create and register our service
+	pgctldService := &PgCtldService{
+		logger: logger,
+	}
+	pb.RegisterPgCtldServer(grpcServer, pgctldService)
+
+	// Create listener
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		return fmt.Errorf("failed to listen on port %d: %w", port, err)
+	}
+
+	logger.Info("Starting pgctld gRPC server", "port", port)
+
+	// Handle graceful shutdown
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	// Start server in goroutine
+	go func() {
+		if err := grpcServer.Serve(lis); err != nil {
+			logger.Error("gRPC server failed", "error", err)
+		}
+	}()
+
+	logger.Info("pgctld gRPC server listening", "address", lis.Addr().String())
+
+	// Wait for shutdown signal
+	<-ctx.Done()
+
+	logger.Info("Shutting down pgctld gRPC server")
+	grpcServer.GracefulStop()
+
+	return nil
+}
+
+// PgCtldService implements the pgctld gRPC service
+type PgCtldService struct {
+	pb.UnimplementedPgCtldServer
+	logger *slog.Logger
+}
+
+func (s *PgCtldService) Start(ctx context.Context, req *pb.StartRequest) (*pb.StartResponse, error) {
+	s.logger.Info("gRPC Start request", "data_dir", req.DataDir, "port", req.Port)
+
+	// Create config from request parameters
+	config := NewPostgresConfigFromStartRequest(req)
+
+	if config.DataDir == "" {
+		return nil, fmt.Errorf("data-dir is required")
+	}
+
+	// Check if PostgreSQL is already running before attempting to start
+	if isPostgreSQLRunning(config.DataDir) {
```

Comment:
It looks like this is symmetric to the implementation of the start command, but with some subtle differences. For example, here we call readPostmasterPID, while in the other one we don’t. Is that intentional? If so, could we add comments explaining how they differ?

For the parts that are the same, what do you think about encapsulating more of the Start functionality so that both server and CLI can share as much as possible? (or another idea, what if from the get-go, the cli just makes the RPC call. So they are always a very thin layer).

Reply:
The difference is in fact subtle. The CLI is idempotent - i.e. if you call start again, it just returns success if a server is already running with those params, but the RPC returns failure. Do you think we should make the RPC idempotent as well?
The point about refactoring is good. I'll take care of that.

Reply:
I have implemented the refactor.